File: examples/Python/bstarsrv.py
# Binary Star Server
#
# Author: Dan Colish <dcolish@gmail.com>

from argparse import ArgumentParser
import time

from zhelpers import zmq

STATE_PRIMARY = 1
STATE_BACKUP = 2
STATE_ACTIVE = 3
STATE_PASSIVE = 4

PEER_PRIMARY = 1
PEER_BACKUP = 2
PEER_ACTIVE = 3
PEER_PASSIVE = 4
CLIENT_REQUEST = 5

HEARTBEAT = 1000


class BStarState(object):
    def __init__(self, state, event, peer_expiry):
        self.state = state
        self.event = event
        self.peer_expiry = peer_expiry


class BStarException(Exception):
    pass

fsm_states = {
    STATE_PRIMARY: {
        PEER_BACKUP: ("I: connected to backup (slave), ready as master",
                      STATE_ACTIVE),
        PEER_ACTIVE: ("I: connected to backup (master), ready as slave",
                      STATE_PASSIVE)
        },
    STATE_BACKUP: {
        PEER_ACTIVE: ("I: connected to primary (master), ready as slave",
                      STATE_PASSIVE),
        CLIENT_REQUEST: ("", False)
        },
    STATE_ACTIVE: {
        PEER_ACTIVE: ("E: fatal error - dual masters, aborting", False)
        },
    STATE_PASSIVE: {
        PEER_PRIMARY: ("I: primary (slave) is restarting, ready as master",
                       STATE_ACTIVE),
        PEER_BACKUP: ("I: backup (slave) is restarting, ready as master",
                      STATE_ACTIVE),
        PEER_PASSIVE: ("E: fatal error - dual slaves, aborting", False),
        CLIENT_REQUEST: (CLIENT_REQUEST, True)  # Say true, check peer later
        }
    }


def run_fsm(fsm):
    # There are some transitional states we do not want to handle
    state_dict = fsm_states.get(fsm.state, {})
    res = state_dict.get(fsm.event)
    if res:
        msg, state = res
    else:
        return
    if state == False:
        raise BStarException(msg)
    elif msg == CLIENT_REQUEST:
        assert fsm.peer_expiry > 0
        if int(time.time() * 1000) > fsm.peer_expiry:
            fsm.state = STATE_ACTIVE
        else:
            raise BStarException()
    else:
        print msg
        fsm.state = state


def main():
    parser = ArgumentParser()
    group = parser.add_mutually_exclusive_group()
    group.add_argument("-p", "--primary", action="store_true", default=False)
    group.add_argument("-b", "--backup", action="store_true", default=False)
    args = parser.parse_args()

    ctx = zmq.Context()
    statepub = ctx.socket(zmq.PUB)
    statesub = ctx.socket(zmq.SUB)
    statesub.setsockopt(zmq.SUBSCRIBE, "")
    frontend = ctx.socket(zmq.ROUTER)

    fsm = BStarState(0, 0, 0)

    if args.primary:
        print "I: Primary master, waiting for backup (slave)"
        frontend.bind("tcp://*:5001")
        statepub.bind("tcp://*:5003")
        statesub.connect("tcp://localhost:5004")
        fsm.state = STATE_PRIMARY
    elif args.backup:
        print "I: Backup slave, waiting for primary (master)"
        frontend.bind("tcp://*:5002")
        statepub.bind("tcp://*:5004")
        statesub.connect("tcp://localhost:5003")
        statesub.setsockopt(zmq.SUBSCRIBE, "")
        fsm.state = STATE_BACKUP

    send_state_at = int(time.time() * 1000 + HEARTBEAT)
    poller = zmq.Poller()
    poller.register(frontend, zmq.POLLIN)
    poller.register(statesub, zmq.POLLIN)

    while True:
        time_left = send_state_at - int(time.time() * 1000)
        if time_left < 0:
            time_left = 0
        socks = dict(poller.poll(time_left))
        if socks.get(frontend) == zmq.POLLIN:
            msg = frontend.recv_multipart()
            fsm.event = CLIENT_REQUEST
            try:
                run_fsm(fsm)
                frontend.send_multipart(msg)
            except BStarException:
                del msg

        if socks.get(statesub) == zmq.POLLIN:
            msg = statesub.recv()
            fsm.event = int(msg)
            del msg
            try:
                run_fsm(fsm)
                fsm.peer_expiry = int(time.time() * 1000) + (2 * HEARTBEAT)
            except BStarException:
                break
        if int(time.time() * 1000) >= send_state_at:
            statepub.send("%d" % fsm.state)
            send_state_at = int(time.time() * 1000) + HEARTBEAT

if __name__ == '__main__':
    main()

File: examples/Ruby/bstarsrv.rb
#!/usr/bin/env ruby
# vim: ft=ruby

# Binary Star server proof-of-concept implementation. This server does no
# real work; it just demonstrates the Binary Star failover model.

require 'optparse'
require 'cztop'

#  We send state information this often
#  If peer doesn't respond in two heartbeats, it is 'dead'
HEARTBEAT = 1000 # in msecs

# Binary Star finite-state machine.
class BStarState
  Exception = Class.new(StandardError)

  attr_accessor :state
  attr_writer :peer_expiry

  def initialize(state, peer_expiry = nil)
    unless [:primary, :backup].include? state
      abort "invalid initial state #{state.inspect}"
    end

    @state = state
    @peer_expiry = peer_expiry
  end

  def <<(event)
    puts "processing event #{event.inspect} ..."
    case @state
    when :primary
      case event
      when :peer_backup
        puts "I: connected to backup (passive), ready active"
        @state = :active
      when :peer_active
        puts "I: connected to backup (active), ready passive"
        @state = :passive
      end
      # Accept client connections

    when :backup
      case event
      when :peer_active
        puts "I: connected to primary (active), ready passive"
        @state = :passive
      when :client_request
        # Reject client connections when acting as backup
        raise Exception, "not active"
      end

    when :active
      case event
      when :peer_active
        # Two actives would mean split-brain
        puts "E: fatal error - dual actives, aborting"
        abort "split brain"
      end

    when :passive
      case event
      when :peer_primary
        # Peer is restarting - become active, peer will go passive
        puts "I: primary (passive) is restarting, ready active"
        @state = :active

      when :peer_backup
        # Peer is restarting - become active, peer will go passive
        puts "I: backup (passive) is restarting, ready active"
        @state = :active;

      when :peer_passive
        # Two passives would mean cluster would be non-responsive
        puts "E: fatal error - dual passives, aborting"
        abort "dual passives"

      when :client_request
        # Peer becomes active if timeout has passed
        # It's the client request that triggers the failover
        abort "bad peer expiry" unless @peer_expiry
        if Time.now >= @peer_expiry
          # If peer is dead, switch to the active state
          puts "I: failover successful, ready active"
          @state = :active
        else
          # If peer is alive, reject connections
          raise Exception, "peer is alive"
        end
      end
    end
  end
end


if __FILE__ == $0
  options = {}
  OptionParser.new do |opts|
    opts.banner = "Usage: #$0 [options]"

    opts.on("-p", "--primary", "run as primary server") do |v|
      options[:role] = :primary
    end

    opts.on("-b", "--backup", "run as backup server") do |v|
      options[:role] = :backup
    end
  end.parse!

  unless options[:role]
    abort "Usage: #$0 { -p | -b }"
  end

  # We use three sockets; one to publish state, one to subscribe to state, and
  # one for client requests/replies.
  statepub = CZTop::Socket::PUB.new
  statesub = CZTop::Socket::SUB.new
  statesub.subscribe
  frontend = CZTop::Socket::ROUTER.new

  # We bind/connect our sockets with our peer and make sure we will get state
  # messages correctly.
  case options[:role]
  when :primary
    puts "I: Primary master, waiting for backup (slave)"
    statepub.bind("tcp://*:5003")
    statesub.connect("tcp://localhost:5004")
    frontend.bind("tcp://*:5001")
    bstar = BStarState.new(:primary)
  when :backup
    puts "I: Backup slave, waiting for primary (master)"
    statepub.bind("tcp://*:5004")
    statesub.connect("tcp://localhost:5003")
    statesub.subscribe
    frontend.bind("tcp://*:5002")
    bstar = BStarState.new(:backup)
  end

  # We now process events on our two input sockets, and process these events
  # one at a time via our finite-state machine. Our "work" for a client
  # request is simply to echo it back:
  poller = CZTop::Poller.new(statesub, frontend)
  send_state_at = Time.now + (HEARTBEAT/1000.0)
  while true
    # round to msec resolution to avoid polling bursts
    time_left = (send_state_at - Time.now).round(3)
    time_left = 0 if time_left < 0
    time_left = (time_left * 1000).to_i # convert to msec

    case poller.simple_wait(time_left)
    when statesub
      # state from peer
      msg = statesub.receive
      puts "received message from statesub: #{msg.to_a.inspect}"
      bstar << :"peer_#{msg[0]}" # this could exit the process
      bstar.peer_expiry = Time.now + 2 * (HEARTBEAT/1000.0)
    when frontend
      # client request
      msg = frontend.receive
      puts "received message from frontend: #{msg.to_a.inspect}"
      begin
        bstar << :client_request
        frontend << msg
      rescue BStarState::Exception
        # We got a client request even though we're passive AND peer is alive.
        # We'll just ignore it.
      end
    end

    # If we timed out, send state to peer.
    if Time.now >= send_state_at
      statepub << bstar.state.to_s
      send_state_at = Time.now + (HEARTBEAT/1000.0)
    end
  end
end
